File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.java
package au.com.shiftyjelly.pocketcasts.repositories.file;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import au.com.shiftyjelly.pocketcasts.models.entity.BaseEpisode;
import au.com.shiftyjelly.pocketcasts.models.entity.PodcastEpisode;
import au.com.shiftyjelly.pocketcasts.models.type.EpisodeStatusEnum;
import au.com.shiftyjelly.pocketcasts.preferences.Settings;
import au.com.shiftyjelly.pocketcasts.repositories.podcast.EpisodeManager;
import au.com.shiftyjelly.pocketcasts.utils.FileUtil;
import au.com.shiftyjelly.pocketcasts.utils.StringUtil;
import au.com.shiftyjelly.pocketcasts.utils.log.LogBuffer;
import dagger.hilt.android.qualifiers.ApplicationContext;
import timber.log.Timber;

@Singleton
public class FileStorage {
	public static final String DIR_CUSTOM_FILES = "custom_episodes";
	public static final String DIR_NETWORK_IMAGES = "network_images";
	public static final String DIR_EPISODES = "podcasts";

	private Settings settings;
	private Context context;
	private final FileStorageKtDelegate delegate;

	@Inject
    public FileStorage(Settings settings, @ApplicationContext  Context context) {
        this.settings = settings;
		this.context = context;
		delegate = new FileStorageKtDelegate(settings, context);
    }

	@NonNull
	public File getPodcastEpisodeFile(BaseEpisode episode) throws StorageException {
		String fileName = episode.getUuid() + episode.getFileExtension();
		File directory;
		if (episode instanceof PodcastEpisode) {
			directory = getPodcastDirectory();
		} else {
			directory = getCloudFilesFolder();
		}

		return new File(directory, fileName);
	}

	@NonNull
	public File getTempPodcastEpisodeFile(BaseEpisode episode) throws StorageException {
		String fileName = episode.getUuid() + episode.getFileExtension();
		return new File(getTempPodcastDirectory(), fileName);
	}

	@Nullable
	public File getCloudFileImage(String uuid) {
    	return delegate.getCloudFileImage(uuid);
	}

    public File getCloudFilesFolder() throws StorageException {
    	return delegate.getCloudFilesFolder();
	}

	public File getOpmlFileFolder() throws StorageException {
		return delegate.getOpmlFileFolder();
	}

	public  File getNetworkImageDirectory() throws StorageException {
		return delegate.getNetworkImageDirectory();
	}
	
	public  File getPodcastDirectory() throws StorageException {
		return delegate.getPodcastDirectory();
	}
	
	public  File getTempPodcastDirectory() throws StorageException {
		return delegate.getTempPodcastDirectory();
	}

	public File getOldTempPodcastDirectory() throws StorageException {
    	return delegate.getOldTempPodcastDirectory();
	}

	public  File getPodcastGroupImageDirectory() throws StorageException {
		return delegate.getPodcastGroupImageDirectory();
	}

	public File getOrCreateCacheDirectory(String name) throws StorageException {
    	return delegate.getOrCreateCacheDirectory(name);
	}

    public File getOrCreateDirectory(String name) throws StorageException {
		return delegate.getOrCreateDirectory(name);
	}
	
	private File getOrCreateDirectory(File parentDir, String name) throws StorageException {
		return delegate.getOrCreateDirectory(parentDir, name);
	}
	
	public File getStorageDirectory() throws StorageException {
		return delegate.getStorageDirectory();
	}
	
	public File getBaseStorageDirectory() throws StorageException {
		return delegate.getBaseStorageDirectory();
	}
	
	public File getBaseStorageDirectory(String choice) throws StorageException {
		return delegate.getBaseStorageDirectory(choice);
	}

	private File createDirectory(File dir) {
		return delegate.createDirectory(dir);
	}

	private void addNoMediaFile(File folder) {
		delegate.addNoMediaFile(folder);
	}
	
	// confirms that all the folders we want to hide from the user have .nomedia files in them
	public void checkNoMediaDirs(){
		// we can do this by getting all the folders
		try {
            addNoMediaFile(getStorageDirectory());
			getNetworkImageDirectory();
			getPodcastGroupImageDirectory();
			getTempPodcastDirectory();
		}
		catch (Exception e) {
			Timber.e(e);
		}
	}

	private String moveFileToDirectory(String filePath, File directory) {
		// validate the path, check PocketCasts is in the path so we don't delete something important
		if (StringUtil.isBlank(filePath) || filePath.indexOf("/PocketCasts") == -1) {
			LogBuffer.INSTANCE.e(LogBuffer.TAG_BACKGROUND_TASKS, "Not moving because it's blank or not PocketCasts");
			return filePath;
		}

		File file = new File(filePath);
        // check we aren't copying to the same directory
        if (file.getParentFile().equals(directory)) {
			LogBuffer.INSTANCE.e(LogBuffer.TAG_BACKGROUND_TASKS, "Not moving because it's the same directory");
            return filePath;
        }
		File newFile = new File(directory, file.getName());
		if (file.exists() && file.isFile()) {
			try {
				FileUtil.copyFile(file, newFile);
				boolean wasDeleted = file.delete();
				LogBuffer.INSTANCE.i(LogBuffer.TAG_BACKGROUND_TASKS, "Moved " + file.getAbsolutePath() + " to " + newFile.getAbsolutePath() + " wasDeleted: " + wasDeleted);
			} catch (IOException e) {
				LogBuffer.INSTANCE.e(LogBuffer.TAG_BACKGROUND_TASKS, e,"Problems moving a file to a new location. from: "+file.getAbsolutePath()+" to: "+newFile.getAbsolutePath());
			}
		}
		return newFile.getAbsolutePath();
	}
	
	private void moveDirectory(File fromDirectory, File toDirectory) {
		if (fromDirectory == null || !fromDirectory.exists() || !fromDirectory.isDirectory() || toDirectory == null) {
			return;
		}

		try {
			FileUtil.copyDirectory(fromDirectory, toDirectory);
			fromDirectory.delete();
		} catch (IOException e) {
			Timber.e(e,"Problems moving a directory to a new location. from: "+fromDirectory.getAbsolutePath()+" to: "+toDirectory.getAbsolutePath());
		}
	}

	public void moveStorage(File oldDirectory, File newDirectory, EpisodeManager episodeManager) {
		try {
			final File pocketCastsDir = new File(oldDirectory, "PocketCasts");
	 	   	if (pocketCastsDir.exists() && pocketCastsDir.isDirectory()) {
				LogBuffer.INSTANCE.i(LogBuffer.TAG_BACKGROUND_TASKS, "Pocket casts directory exists");
				
	 	   		newDirectory.mkdirs();
	 	   		final File newPocketCastsDir = new File(newDirectory, "PocketCasts");

				boolean folderAlreadyExisted = newPocketCastsDir.exists();

				// check existing media and mark those episodes as downloaded
				final File episodeDirectory = getOrCreateDirectory(newPocketCastsDir, DIR_EPISODES);
				if (folderAlreadyExisted) {
					boolean foundMedia = false;
					File[] files = episodeDirectory.listFiles();
					if (files != null) {
						for (File file : files) {
							String fileName = FileUtil.getFileNameWithoutExtension(file);
							if (fileName.length() < 36) {
								continue;
							}
							//noinspection deprecation
							PodcastEpisode episode = episodeManager.findByUuidSync(fileName);
							if (episode != null) {
								// Delete the original file if it is already there
								if (!StringUtil.isBlank(episode.getDownloadedFilePath())) {
									File originalFile = new File(episode.getDownloadedFilePath());
									if (originalFile.exists()) {
										originalFile.delete();
									}
								}
								episodeManager.updateDownloadFilePath(episode, file.getAbsolutePath(), true);

								foundMedia = true;
							}
						}
					}
				}

				// move episodes
				List<PodcastEpisode> episodes = episodeManager.observeDownloadedEpisodes().blockingFirst();
				for (PodcastEpisode episode : episodes) {
					LogBuffer.INSTANCE.i(LogBuffer.TAG_BACKGROUND_TASKS, "Found downloaded episode " + episode.getTitle());
					String episodeFilePath = episode.getDownloadedFilePath();
					if (StringUtil.isBlank(episodeFilePath)) {
						LogBuffer.INSTANCE.e(LogBuffer.TAG_BACKGROUND_TASKS, "Episode had no file path");
						continue;
					}
					File file = new File(episodeFilePath);
					if (file.exists() && file.isFile()) {
						episode.setDownloadedFilePath(moveFileToDirectory(episodeFilePath, episodeDirectory));
						episodeManager.updateDownloadFilePath(episode, episode.getDownloadedFilePath(), false);
					}
				}

                // move custom folders
                File customFilesDirectory = getOrCreateDirectory(pocketCastsDir, DIR_CUSTOM_FILES);
                File newCustomFilesDirectory = getOrCreateDirectory(newPocketCastsDir, DIR_CUSTOM_FILES);
                if (customFilesDirectory.exists()) {
                    moveDirectory(customFilesDirectory, newCustomFilesDirectory);
                }
	 	   	    
	 	   	    // move network and group images
		 	   	File networkImageDirectory = getOrCreateDirectory(pocketCastsDir, DIR_NETWORK_IMAGES);
	 	   	    File newNetworkImageDirectory = getOrCreateDirectory(newPocketCastsDir, DIR_NETWORK_IMAGES);
	 	   	    if (networkImageDirectory.exists()) {
	 	   	    	moveDirectory(networkImageDirectory, newNetworkImageDirectory);
	 	   	    }
	 	   	} else {
	 	   		LogBuffer.INSTANCE.e(LogBuffer.TAG_BACKGROUND_TASKS, "Old directory did not exist");
			}
		}
		catch(StorageException e) {
			LogBuffer.INSTANCE.e(LogBuffer.TAG_BACKGROUND_TASKS, e,"Unable to move storage to new location.");
		}
	}

	public void fixBrokenFiles(EpisodeManager episodeManager) {
		try {
			// get all possible locations
			List<String> folderPaths = new ArrayList<>();
			for (FolderLocation location : new StorageOptions().getFolderLocations(context)) {
				folderPaths.add(location.getFilePath());
			}
			folderPaths.add(context.getFilesDir().getAbsolutePath());
			String customFolder = settings.getStorageCustomFolder();
			if (StringUtil.isPresent(customFolder) && !folderPaths.contains(customFolder) && new File(customFolder).exists()) {
				folderPaths.add(customFolder);
			}

			// search each folder for missing files
			for (String folderPath : folderPaths) {
				File folder = new File(folderPath);
				if (!folder.exists() || !folder.canRead()) {
					continue;
				}
				File pocketcastsFolder = new File(folder, "PocketCasts");
				if (!pocketcastsFolder.exists() || !pocketcastsFolder.canRead()) {
					continue;
				}
				File episodesFolder = new File(pocketcastsFolder, DIR_EPISODES);
				if (!episodesFolder.exists() || !episodesFolder.canRead()) {
					continue;
				}
				File[] files = episodesFolder.listFiles();
				for (int i=0; i<files.length; i++) {
					File file = files[i];
					String filename = file.getName();
					int dotPosition = filename.lastIndexOf(".");
					if (dotPosition < 1) {
						continue;
					}
					String uuid = filename.substring(0, dotPosition);
					if (uuid.length() != 36) {
						continue;
					}
					//noinspection deprecation
					PodcastEpisode episode = episodeManager.findByUuidSync(uuid);
					if (episode != null) {
						if (episode.getDownloadedFilePath() != null && new File(episode.getDownloadedFilePath()).exists() && episode.isDownloaded()) {
							// skip as the episode download already exists
							continue;
						}

						LogBuffer.INSTANCE.i(LogBuffer.TAG_BACKGROUND_TASKS, "Restoring downloaded file for " + episode.getTitle() + " from " + file.getAbsolutePath());
						// link to the found episode
						episode.setEpisodeStatus(EpisodeStatusEnum.DOWNLOADED);
						episode.setDownloadedFilePath(file.getAbsolutePath());
						episodeManager.update(episode);
					}
				}
			}
		}
		catch (Exception e) {
			Timber.e(e);
		}
	}
}

File: modules/services/repositories/src/main/java/au/com/shiftyjelly/pocketcasts/repositories/file/FileStorage.kt
package au.com.shiftyjelly.pocketcasts.repositories.file

import android.content.Context
import au.com.shiftyjelly.pocketcasts.models.entity.BaseEpisode
import au.com.shiftyjelly.pocketcasts.models.entity.PodcastEpisode
import au.com.shiftyjelly.pocketcasts.models.entity.UserEpisode
import au.com.shiftyjelly.pocketcasts.models.type.EpisodeStatusEnum
import au.com.shiftyjelly.pocketcasts.preferences.Settings
import au.com.shiftyjelly.pocketcasts.repositories.podcast.EpisodeManager
import au.com.shiftyjelly.pocketcasts.utils.FileUtil
import au.com.shiftyjelly.pocketcasts.utils.log.LogBuffer
import dagger.hilt.android.qualifiers.ApplicationContext
import java.io.File
import java.io.IOException
import javax.inject.Inject
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.reactive.asFlow
import timber.log.Timber

open class FileStorage @Inject constructor(
    val settings: Settings,
    @ApplicationContext val context: Context,
) {
    fun getOrCreatePodcastEpisodeFile(episode: BaseEpisode): File? {
        val fileName = episode.uuid + episode.getFileExtension()
        val episodeDir = when (episode) {
            is PodcastEpisode -> getOrCreateEpisodesDir()
            is UserEpisode -> getOrCreateCloudDir()
        }
        return episodeDir?.let { dir -> File(dir, fileName) }
    }

    fun getOrCreatePodcastEpisodeTempFile(episode: BaseEpisode): File {
        val fileName = episode.uuid + episode.getFileExtension()
        return File(getOrCreateEpisodesTempDir(), fileName)
    }

    fun getOrCreateCloudFileImage(uuid: String): File? = try {
        val fileName = uuid + "_imagefile"
        getOrCreateCloudDir()?.let { dir -> File(dir, fileName) }
    } catch (e: StorageException) {
        Timber.e(e)
        null
    }

    fun getOrCreateCloudDir(): File? = getOrCreateDir(DIR_CLOUD_FILES)

    fun getOrCreateOpmlDir(): File? = getOrCreateDir(DIR_OPML_FILES)

    fun getOrCreateNetworkImagesDir(): File? = getOrCreateDir(DIR_NETWORK_IMAGES)

    fun getOrCreateEpisodesDir(): File? = getOrCreateDir(DIR_EPISODES)

    fun getOrCreateEpisodesTempDir(): File = getOrCreateCacheDir(DIR_TEMP_EPISODES)

    fun getOrCreateEpisodesOldTempDir(): File? = getOrCreateDir(DIR_TEMP_EPISODES)

    fun getOrCreatePodcastGroupImagesDir(): File? = getOrCreateDir(DIR_PODCAST_GROUP_IMAGES)

    fun getOrCreateCacheDir(name: String): File = getOrCreateDir(context.cacheDir, name)

    fun getOrCreateDir(name: String): File? = getOrCreateStorageDir()?.let { dir ->
        getOrCreateDir(dir, name)
    }

    private fun getOrCreateDir(parentDir: File, name: String): File = File(parentDir, name + File.separator).also { dir ->
        createDir(dir)
        addNoMediaFile(dir)
    }

    fun getOrCreateStorageDir(): File? = getOrCreateBaseStorageDir()?.let { dir ->
        File(dir, "PocketCasts" + File.separator).also(::createDir)
    }

    fun getOrCreateBaseStorageDir(): File? = settings.getStorageChoice()?.let(::getOrCreateBaseStorageDir)

    private fun getOrCreateBaseStorageDir(choice: String): File = if (choice == Settings.STORAGE_ON_CUSTOM_FOLDER) {
        val path = settings.getStorageCustomFolder()
        if (path.isBlank()) {
            throw StorageException("Ooops, please set the Custom Folder Location in the settings.")
        }
        val storageCustomDir = File(path)
        if (!storageCustomDir.exists() && !storageCustomDir.mkdirs()) {
            throw StorageException("Storage custom folder unavailable.")
        }
        storageCustomDir
    } else {
        File(choice)
    }

    private fun createDir(dir: File): File = dir.also(File::mkdirs)

    private fun addNoMediaFile(dir: File) {
        if (!dir.exists()) {
            return
        }
        val noMediaFile = File(dir, ".nomedia")
        if (!noMediaFile.exists()) {
            try {
                noMediaFile.createNewFile()
            } catch (e: Exception) {
                Timber.e(e)
            }
        }
    }

    /**
     * Confirms that all the directories we want to hide from the user have .nomedia files in them
     */
    fun checkNoMediaDirs() {
        // Getting a directory also adds .nomedia file to that dir
        try {
            getOrCreateStorageDir()?.let(::addNoMediaFile)
            getOrCreateNetworkImagesDir()
            getOrCreatePodcastGroupImagesDir()
            getOrCreateEpisodesTempDir()
        } catch (e: Exception) {
            Timber.e(e)
        }
    }

    private fun moveFileToDir(filePath: String, dir: File): String? {
        // Validate the path, check PocketCasts is in the path so we don't delete something important
        if (filePath.isBlank() || "/PocketCasts" !in filePath) {
            LogBuffer.e(LogBuffer.TAG_BACKGROUND_TASKS, "Not moving because it's blank or not PocketCasts")
            return filePath
        }

        val file = File(filePath)
        // Check we aren't copying to the same directory
        if (file.parentFile == dir) {
            LogBuffer.e(LogBuffer.TAG_BACKGROUND_TASKS, "Not moving because it's the same directory")
            return filePath
        }

        val newFile = File(dir, file.name)
        if (file.exists() && file.isFile) {
            try {
                FileUtil.copyFile(file, newFile)
                val wasDeleted = file.delete()
                LogBuffer.i(LogBuffer.TAG_BACKGROUND_TASKS, "Moved ${file.absolutePath} to ${newFile.absolutePath} wasDeleted: $wasDeleted")
            } catch (e: IOException) {
                LogBuffer.e(LogBuffer.TAG_BACKGROUND_TASKS, e, "Problems moving a file to a new location. from: ${file.absolutePath} to: ${newFile.absolutePath}")
            }
        }

        return newFile.absolutePath
    }

    private fun moveDir(fromDir: File, toDir: File) {
        if (fromDir.exists() && fromDir.isDirectory) {
            try {
                FileUtil.copyDirectory(fromDir, toDir)
                fromDir.delete()
            } catch (e: IOException) {
                Timber.e(e, "Problems moving a  directory to a new location. from: ${fromDir.absolutePath} to: ${toDir.absolutePath}")
            }
        }
    }

    suspend fun moveStorage(oldDir: File, newDir: File, episodesManager: EpisodeManager) {
        try {
            val oldPocketCastsDir = File(oldDir, "PocketCasts")
            if (oldPocketCastsDir.exists() && oldPocketCastsDir.isDirectory) {
                LogBuffer.i(LogBuffer.TAG_BACKGROUND_TASKS, "Pocket casts directory exists")

                newDir.mkdirs()
                val newPocketCastsDir = File(newDir, "PocketCasts")
                val episodesDir = getOrCreateDir(newPocketCastsDir, DIR_EPISODES)

                // Check existing media and mark those episodes as downloaded
                episodesDir.takeIf(File::exists)?.listFiles().orEmpty()
                    .matchWithFileNames()
                    .filterInvalidFileNames()
                    .findMatchingEpisodes(episodesManager)
                    .deleteExistingFiles()
                    .updateEpisodesWithNewFilePaths(episodesManager)

                // Move episodes
                episodesManager.observeDownloadedEpisodes().asFlow().first()
                    .onEach { episode -> LogBuffer.i(LogBuffer.TAG_BACKGROUND_TASKS, "Found downloaded episode ${episode.title}") }
                    .matchWithDownloadedFilePaths()
                    .filterNotExistingFiles()
                    .moveFilesToEpisodesDirAndUpdatePaths(episodesManager, episodesDir)

                val oldCustomFilesDir = getOrCreateDir(oldPocketCastsDir, DIR_CUSTOM_EPISODES)
                val newCustomFilesDir = getOrCreateDir(newPocketCastsDir, DIR_CUSTOM_EPISODES)
                moveDir(oldCustomFilesDir, newCustomFilesDir)

                val oldNetworkImageDir = getOrCreateDir(oldPocketCastsDir, DIR_NETWORK_IMAGES)
                val newNetworkImageDir = getOrCreateDir(newPocketCastsDir, DIR_NETWORK_IMAGES)
                moveDir(oldNetworkImageDir, newNetworkImageDir)
            } else {
                LogBuffer.e(LogBuffer.TAG_BACKGROUND_TASKS, "Old directory did not exist")
            }
        } catch (e: StorageException) {
            LogBuffer.e(LogBuffer.TAG_BACKGROUND_TASKS, e, "Unable to move storage to new location")
        }
    }

    private fun Array<out File>.matchWithFileNames() = map { file -> file to FileUtil.getFileNameWithoutExtension(file) }

    private fun List<Pair<File, String>>.filterInvalidFileNames() = filter { (_, fileName) -> fileName.length >= UUID_LENGTH }

    private suspend fun List<Pair<File, String>>.findMatchingEpisodes(episodeManager: EpisodeManager) = mapNotNull { (file, fileName) ->
        episodeManager.findByUuid(fileName)?.let { episode -> file to episode }
    }

    private fun List<Pair<File, PodcastEpisode>>.deleteExistingFiles() = onEach { (_, episode) ->
        episode.downloadedFilePath?.takeIf(String::isNotBlank)?.let { downloadedFilePath ->
            val originalFile = File(downloadedFilePath)
            if (originalFile.exists()) {
                originalFile.delete()
            }
        }
    }

    private fun List<Pair<File, PodcastEpisode>>.updateEpisodesWithNewFilePaths(episodeManager: EpisodeManager) = forEach { (file, episode) ->
        episodeManager.updateDownloadFilePath(episode, file.absolutePath, markAsDownloaded = true)
    }

    private fun List<PodcastEpisode>.matchWithDownloadedFilePaths() = mapNotNull { episode ->
        val downloadedFilePath = episode.downloadedFilePath?.takeIf(String::isNotBlank)
        if (downloadedFilePath == null) {
            LogBuffer.e(LogBuffer.TAG_BACKGROUND_TASKS, "Episode had not file path")
        }
        downloadedFilePath?.let { path -> episode to path }
    }

    private fun List<Pair<PodcastEpisode, String>>.filterNotExistingFiles() = filter { (_, path) -> File(path).takeIf { it.exists() && it.isFile } != null }

    private fun List<Pair<PodcastEpisode, String>>.moveFilesToEpisodesDirAndUpdatePaths(episodeManager: EpisodeManager, episodesDir: File) = forEach { (episode, path) ->
        moveFileToDir(path, episodesDir)?.let { updatedPath ->
            episodeManager.updateDownloadFilePath(episode, updatedPath, markAsDownloaded = false)
        }
    }

    suspend fun fixBrokenFiles(episodeManager: EpisodeManager) {
        try {
            // Get all possible locations
            val dirPaths = buildSet {
                addAll(StorageOptions().getFolderLocations(context).map(FolderLocation::filePath))
                add(context.filesDir.absolutePath)
                val customDir = settings.getStorageCustomFolder()
                if (customDir.isNotBlank() && File(customDir).exists()) {
                    add(customDir)
                }
            }

            dirPaths.asSequence()
                .toExistingDirs()
                .toExistingPocketCastsDirs()
                .toExistingEpisodesDirs()
                .listFiles()
                .matchWithFileNameDotPosition()
                .toFileNameUuids()
                .asFlow()
                .toMatchingEpisodes(episodeManager)
                .filterDownloadedEpisodes()
                .restoreDownloadedFilePaths(episodeManager)
        } catch (e: Exception) {
            Timber.e(e)
        }
    }

    private fun Sequence<String>.toExistingDirs() = mapNotNull { dirPath -> File(dirPath).takeIf(::isFileReadable) }

    private fun Sequence<File>.toExistingPocketCastsDirs() = mapNotNull { parentDir -> File(parentDir, "PocketCasts").takeIf(::isFileReadable) }

    private fun Sequence<File>.toExistingEpisodesDirs() = mapNotNull { parentDir -> File(parentDir, DIR_EPISODES).takeIf(::isFileReadable) }

    private fun isFileReadable(file: File) = file.exists() && file.canRead()

    private fun Sequence<File>.listFiles() = flatMap { it.listFiles().orEmpty().asSequence() }

    private fun Sequence<File>.matchWithFileNameDotPosition() = mapNotNull { file ->
        file.name.lastIndexOf('.').takeIf { it >= 1 }?.let { dotPosition -> file to dotPosition }
    }

    private fun Sequence<Pair<File, Int>>.toFileNameUuids() = mapNotNull { (file, dotPosition) ->
        file.name.substring(0, dotPosition).takeIf { it.length != UUID_LENGTH }?.let { uuid -> file to uuid }
    }

    private fun Flow<Pair<File, String>>.toMatchingEpisodes(episodeManager: EpisodeManager) = mapNotNull { (file, uuid) ->
        episodeManager.findByUuid(uuid)?.let { episode -> file to episode }
    }

    private fun Flow<Pair<File, PodcastEpisode>>.filterDownloadedEpisodes() = filter { (_, episode) ->
        val downloadedFilePath = episode.downloadedFilePath
        downloadedFilePath == null || !File(downloadedFilePath).exists() || !episode.isDownloaded
    }

    private suspend fun Flow<Pair<File, PodcastEpisode>>.restoreDownloadedFilePaths(episodeManager: EpisodeManager) = collect { (file, episode) ->
        LogBuffer.i(LogBuffer.TAG_BACKGROUND_TASKS, "Restoring downloaded file for ${episode.title} from ${file.absolutePath}")
        // Link to the found episode
        episode.episodeStatus = EpisodeStatusEnum.DOWNLOADED
        episode.downloadedFilePath = file.absolutePath
        episodeManager.update(episode)
    }

    private companion object {
        const val DIR_EPISODES = "podcasts"
        const val DIR_CUSTOM_EPISODES = "custom_episodes"
        const val DIR_TEMP_EPISODES = "downloadTmp"
        const val DIR_NETWORK_IMAGES = "network_images"
        val DIR_PODCAST_GROUP_IMAGES = "network_images" + File.separator + "groups" + File.separator
        const val DIR_CLOUD_FILES = "cloud_files"
        const val DIR_OPML_FILES = "opml_import"

        const val UUID_LENGTH = 36
    }
}
